src/QueryPipeline/QueryPipeline.cpp

```diff
@@ -572,7 +571,25 @@ Block QueryPipeline::getHeader() const
 
 void QueryPipeline::setProgressCallback(const ProgressCallback & callback)
 {
-    progress_callback = callback;
+    progress_callback = [callback](const Progress & progress)
+    {
+        // Performance counters need to be updated from the same thread the query is being executed
+        // on because most info is taken using getrusage with RUSAGE_THREAD. Ideally, we would only
+        // update the counters once we're close to the interval at which the query metric log data
+        // needs to be collected. However, since the progress callback is called not very
+        // frequently, we'd rather update them as needed. Using the
+        // updatePerformanceCountersIfNeeded instead of just updatePerformanceCounters we make sure
+        // that we don't update them too frequently.
+        auto context = CurrentThread::getQueryContext();
+        if (context)
+        {
+            if (auto query_metric_log = context->getQueryMetricLog())
+                CurrentThread::updatePerformanceCountersIfNeeded();
```

Comment:
I'm guessing this will be quite inexact, but not more than the current counters we already have for queries and so on.

Reply:
That's also my assumption